nvim client compatibility fails with null/empty config data

Started a basic `nvim-lspconfig` config (please ignore all my horrible assumptions)

```lua
local fs = require('vim.fs')

return {
  default_config = {
    cmd = { 'java', '-jar', fs.abspath('~/vscode/natls/libs/natls/build/libs/natls.jar') },
    filetypes = { 'natural' },
    root_dir = function(dir)
      return fs.root(dir, '.natural')
    end,
  },
  docs = {
    description = [[
https://github.com/MarkusAmshove/vscode-natural

Natural Language Server
]],
  },
}
```

Sadly on connection it stops on this error

```
Error executing vim.schedule lua callback: /usr/local/share/nvim/runtime/lua/vim/lsp/client.lua:548: RPC[Error] code_name = InternalError, message = "Internal error." data = 'java.util.concurrent.CompletionException: java.lang.NullPointerException: Cannot invoke "org.amshove.natls.config.InitializationConfiguration.isAsync()" because the return value of "org.amshove.natls.config.LSConfiguration.getInitialization()" is null\n\tat java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:315)\n\tat java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:320)\n\tat java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1770)\n\tat java.base/java.util.concurrent.CompletableFuture$AsyncSupply.exec(CompletableFuture.java:1760)\n\tat java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)\n\tat java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312)\n\tat java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1843)\n\tat java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808)\n\tat java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188)\nCaused by: java.lang.NullPointerException: Cannot invoke "org.amshove.natls.config.InitializationConfiguration.isAsync()" because the return value of "org.amshove.natls.config.LSConfiguration.getInitialization()" is null\n\tat org.amshove.natls.languageserver.NaturalLanguageService.indexProject(NaturalLanguageService.java:115)\n\tat org.amshove.natls.languageserver.NaturalLanguageServer.lambda$initialize$1(NaturalLanguageServer.java:136)\n\tat java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1768)\n\t... 6 more\n'
stack traceback:
	[C]: in function 'assert'
	/usr/local/share/nvim/runtime/lua/vim/lsp/client.lua:548: in function ''
	vim/_editor.lua: in function <vim/_editor.lua:0>
```

Not sure which path it's taking to initialize the config. I tried a couple of configs passing a struct to `init_options` but couldn't manage to set the `async` property.

A small patch to `NaturalLanguageService` on line 115 ...

```java
		if (getConfig().getInitialization() != null && !getConfig().getInitialization().isAsync())
```

... and `nvim` works (not had a proper play with it yet).

![Image](https://github.com/user-attachments/assets/2cf9c7d8-bc33-429d-88bc-614e0d979127)

It seems to be [generally agreed](https://github.com/microsoft/language-server-protocol/issues/567) that starting a server without initialization options should "just work", but I'm not sure this is what's happening ; wondering if nvim's LSP client is supplying an empty string value here.

--- a/libs/natls/src/main/java/org/amshove/natls/config/CompletionConfiguration.java
+++ b/libs/natls/src/main/java/org/amshove/natls/config/CompletionConfiguration.java
@@ -2,7 +2,7 @@ package org.amshove.natls.config;
 
 public class CompletionConfiguration
 {
-	private boolean qualify;
+	private boolean qualify = false;
 
 	public boolean isQualify()
 	{

--- a/libs/natls/src/main/java/org/amshove/natls/config/InitilizationConfiguration.java
+++ b/libs/natls/src/main/java/org/amshove/natls/config/InitilizationConfiguration.java
@@ -2,7 +2,7 @@ package org.amshove.natls.config;
 
 public class InitilizationConfiguration
 {
-	private boolean async;
+	private boolean async = false;
 
 	public boolean isAsync()
 	{

--- a/libs/natls/src/main/java/org/amshove/natls/config/InlayHintsConfiguration.java
+++ b/libs/natls/src/main/java/org/amshove/natls/config/InlayHintsConfiguration.java
@@ -2,7 +2,7 @@ package org.amshove.natls.config;
 
 public class InlayHintsConfiguration
 {
-	private boolean showAssignmentTargetType;
+	private boolean showAssignmentTargetType = false;
 
 	public boolean isShowAssignmentTargetType()
 	{

--- a/libs/natls/src/main/java/org/amshove/natls/config/LSConfiguration.java
+++ b/libs/natls/src/main/java/org/amshove/natls/config/LSConfiguration.java
@@ -1,7 +1,5 @@
 package org.amshove.natls.config;
 
-import org.eclipse.lsp4j.InitializeParams;
-
 public class LSConfiguration
 {
 

--- a/libs/natls/src/main/java/org/amshove/natls/config/LSConfiguration.java
+++ b/libs/natls/src/main/java/org/amshove/natls/config/LSConfiguration.java
@@ -1,5 +1,7 @@
 package org.amshove.natls.config;
 
+import org.eclipse.lsp4j.InitializeParams;
+
 public class LSConfiguration
 {
 

--- a/libs/natls/src/main/java/org/amshove/natls/config/LSConfiguration.java
+++ b/libs/natls/src/main/java/org/amshove/natls/config/LSConfiguration.java
@@ -3,32 +3,14 @@ package org.amshove.natls.config;
 public class LSConfiguration
 {
 
-	private CompletionConfiguration completion;
-	private InlayHintsConfiguration inlayhints;
-	private InitilizationConfiguration initialization;
-	private MapsConfiguration maps;
+	private CompletionConfiguration completion = new CompletionConfiguration();
+	private InlayHintsConfiguration inlayhints = new InlayHintsConfiguration();
+	private InitilizationConfiguration initialization = new InitilizationConfiguration();
+	private MapsConfiguration maps = new MapsConfiguration();
 
 	public static LSConfiguration createDefault()
 	{
-		var config = new LSConfiguration();
-
-		var completion = new CompletionConfiguration();
-		completion.setQualify(false);
-		config.setCompletion(completion);
-
-		var inlay = new InlayHintsConfiguration();
-		inlay.setShowAssignmentTargetType(false);
-		config.setInlayhints(inlay);
-
-		var init = new InitilizationConfiguration();
-		init.setAsync(false);
-		config.setInitialization(init);
-
-		var maps = new MapsConfiguration();
-		maps.setEnablePreview(false);
-		config.setMaps(maps);
-
-		return config;
+		return new LSConfiguration();
 	}
 
 	public CompletionConfiguration getCompletion()

--- a/libs/natls/src/main/java/org/amshove/natls/config/MapsConfiguration.java
+++ b/libs/natls/src/main/java/org/amshove/natls/config/MapsConfiguration.java
@@ -2,7 +2,7 @@ package org.amshove.natls.config;
 
 public class MapsConfiguration
 {
-	private boolean enablePreview;
+	private boolean enablePreview = false;
 
 	public boolean isPreviewEnabled()
 	{

--- a/libs/natls/src/main/java/org/amshove/natls/languageserver/NaturalLanguageServer.java
+++ b/libs/natls/src/main/java/org/amshove/natls/languageserver/NaturalLanguageServer.java
@@ -47,9 +47,7 @@ public class NaturalLanguageServer implements LanguageServer, LanguageClientAwar
 			log.info("Starting initialization");
 			var capabilities = new ServerCapabilities();
 
-			var config = params.getInitializationOptions() != null
-				? new Gson().fromJson((JsonObject) params.getInitializationOptions(), LSConfiguration.class)
-				: LSConfiguration.createDefault();
+			var config = getInitialConfiguration(params);
 			NaturalLanguageService.setConfiguration(config);
 
 			capabilities.setWorkspaceSymbolProvider(true);
@@ -156,6 +154,20 @@ public class NaturalLanguageServer implements LanguageServer, LanguageClientAwar
 			log.info("Initialization done. Took %dms".formatted(initEnd - initStart));
 			return new InitializeResult(capabilities, new ServerInfo(lspName != null ? lspName : "natls", lspVersion != null ? lspVersion : "dev"));
 		});
+	}
+
+	private LSConfiguration getInitialConfiguration(InitializeParams params)
+	{
+		// Some clients send an empty json object by default instead of null when no options
+		// are defined.
+		if (params.getInitializationOptions()instanceof JsonObject obj && obj.isEmpty())
+		{
+			return LSConfiguration.createDefault();
+		}
+
+		return params.getInitializationOptions() != null
+			? new Gson().fromJson((JsonObject) params.getInitializationOptions(), LSConfiguration.class)
+			: LSConfiguration.createDefault();
 	}
 
 	@Override

--- a/libs/natls/src/test/java/org/amshove/natls/initialization/LanguageServerInitializationShould.java
+++ b/libs/natls/src/test/java/org/amshove/natls/initialization/LanguageServerInitializationShould.java
@@ -1,10 +1,18 @@
 package org.amshove.natls.initialization;
 
-import org.amshove.natls.testlifecycle.LanguageServerTest;
-import org.amshove.natls.testlifecycle.LspProjectName;
-import org.amshove.natls.testlifecycle.LspTestContext;
+import com.google.gson.JsonObject;
+import org.amshove.natls.languageserver.NaturalLanguageServer;
+import org.amshove.natls.languageserver.NaturalLanguageService;
+import org.amshove.natls.testlifecycle.*;
+import org.eclipse.lsp4j.InitializeParams;
+import org.eclipse.lsp4j.WorkspaceFolder;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
+
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
 
 import static org.assertj.core.api.AssertionsForInterfaceTypes.assertThat;
 
@@ -41,6 +49,31 @@ class LanguageServerInitializationShould extends LanguageServerTest
 	{
 		assertThat(getContext().getClient().getShownMessages())
 			.anyMatch(m -> m.contains("Parsing references LIBONE."));
+	}
+
+	@Test
+	void useDefaultLSConfigurationWhenEmptyInitOptionsArePassed()
+		throws ExecutionException, InterruptedException, TimeoutException
+	{
+		var params = new InitializeParams();
+		params.setCapabilities(LspProjectNameResolver.createCapabilities());
+		var projectUri = context.project().rootPath().toUri().toString();
+		params.setRootUri(projectUri);
+		params.setWorkspaceFolders(
+			List.of(new WorkspaceFolder(projectUri, "Natural"))
+		);
+
+		var emptyInitOptions = new JsonObject();
+		params.setInitializationOptions(emptyInitOptions);
+
+		var server = new NaturalLanguageServer();
+		server.connect(new StubClient());
+		server.initialize(params).get(1, TimeUnit.MINUTES);
+
+		var effectiveConfig = NaturalLanguageService.getConfig();
+		assertThat(effectiveConfig.getCompletion())
+			.as("Initial configuration was not set correctly")
+			.isNotNull();
 	}
 
 	@Override

--- a/libs/natls/src/test/java/org/amshove/natls/initialization/LanguageServerInitializationShould.java
+++ b/libs/natls/src/test/java/org/amshove/natls/initialization/LanguageServerInitializationShould.java
@@ -1,5 +1,6 @@
 package org.amshove.natls.initialization;
 
+import com.google.gson.Gson;
 import com.google.gson.JsonObject;
 import org.amshove.natls.languageserver.NaturalLanguageServer;
 import org.amshove.natls.languageserver.NaturalLanguageService;
@@ -16,6 +17,7 @@ import java.util.concurrent.TimeoutException;
 
 import static org.assertj.core.api.AssertionsForInterfaceTypes.assertThat;
 
+@SuppressWarnings("deprecation") // Deprecated LSP APIs
 class LanguageServerInitializationShould extends LanguageServerTest
 {
 	private LspTestContext context;
@@ -74,6 +76,36 @@ class LanguageServerInitializationShould extends LanguageServerTest
 		assertThat(effectiveConfig.getCompletion())
 			.as("Initial configuration was not set correctly")
 			.isNotNull();
+	}
+
+	@Test
+	void useDefaultConfigurationValuesWhenOnlyPartialConfigurationIsSupplied()
+		throws ExecutionException, InterruptedException, TimeoutException
+	{
+		var params = new InitializeParams();
+		params.setCapabilities(LspProjectNameResolver.createCapabilities());
+		var projectUri = context.project().rootPath().toUri().toString();
+		params.setRootUri(projectUri);
+		params.setWorkspaceFolders(
+			List.of(new WorkspaceFolder(projectUri, "Natural"))
+		);
+
+		var jsonConfig = "{ \"completion\": { \"qualify\": true } }";
+		params.setInitializationOptions(new Gson().fromJson(jsonConfig, JsonObject.class));
+
+		var server = new NaturalLanguageServer();
+		server.connect(new StubClient());
+		server.initialize(params).get(1, TimeUnit.MINUTES);
+
+		var effectiveConfig = NaturalLanguageService.getConfig();
+
+		assertThat(effectiveConfig.getCompletion().isQualify())
+			.as("User set configuration should be set")
+			.isTrue();
+
+		assertThat(effectiveConfig.getInitialization().isAsync())
+			.as("Configuration not included in partial configuration should have default value")
+			.isFalse();
 	}
 
 	@Override

--- a/libs/natls/src/test/java/org/amshove/natls/testlifecycle/LspProjectNameResolver.java
+++ b/libs/natls/src/test/java/org/amshove/natls/testlifecycle/LspProjectNameResolver.java
@@ -66,7 +66,7 @@ public class LspProjectNameResolver implements ParameterResolver
 		}
 	}
 
-	private ClientCapabilities createCapabilities()
+	public static ClientCapabilities createCapabilities()
 	{
 		var capabilities = new ClientCapabilities();
 		var window = new WindowClientCapabilities();

